### PR TITLE
Update API surface for relationship APIs

### PR DESCRIPTION
**Pull request by abhipsaMisra**
For APIs returning json objects, we will expose 2 variations:
1. `public Mono<DigitalTwinsResponse<String>> createDigitalTwinWithResponse(String digitalTwinId, String digitalTwin) { }`
2. `public <T> Mono<DigitalTwinsResponse<T>> createDigitalTwinWithResponse(String digitalTwinId, Object digitalTwin, Class<T> klazz) { }`

The relationships [API design document](https://github.com/Azure/azure-sdk-for-java/blob/feature/adt/abmisr/updateApiRelationships/sdk/digitaltwins/azure-digitaltwins-core/API%20design.md#relationships) has been updated to reflect this change.

I have also added the following classes:
1. `DigitalTwinsStringSerializer` - handles the custom string serialization logic for sending json strings as-is.
2. `BasicDigitalTwin`, `DigitalTwinMetadata` and `WritableProperty` - helper classes to help serialize the returned digital twin. Referenced in samples. (same as in C#).
3. `DigitalTwinsResponse<T>` and `DigitalTwinsResponseHeaders` - Digital twin specific response types inherited from `ResponseBase<H, T>` for returning digital twin response. It is of generic type, since some return types from public APIs will be `String`, while some will be `<T>`.

### CODE REVIEW

File: sdk/digitaltwins/azure-digitaltwins-core/src/main/java/com/azure/digitaltwins/core/DigitalTwinsAsyncClient.java [hunk 1]
```diff
@@ -149,12 +146,59 @@ public HttpPipeline getHttpPipeline() {
...
      * @param relationshipName The name of a relationship to filter to.
      * @return A {@link PagedFlux} of application/json relationships belonging to the specified digital twin and the http response.
-     * TODO: Impl here returns an Object and not a String.
      */
     @ServiceMethod(returns = ReturnType.COLLECTION)
-    public PagedFlux<Object> listRelationships(String digitalTwinId, String relationshipName) {
-        return new PagedFlux<>(
-            () -> protocolLayer.getDigitalTwins().listRelationshipsSinglePageAsync(digitalTwinId, relationshipName),
-            nextLink -> protocolLayer.getDigitalTwins().listRelationshipsNextSinglePageAsync(nextLink));
+    public PagedFlux<String> listRelationships(String digitalTwinId, String relationshipName) {
```
Comment (abhipsaMisra) [inline sdk/digitaltwins/azure-digitaltwins-core/src/main/java/com/azure/digitaltwins/core/DigitalTwinsAsyncClient.java:151]: The final relationship APIs have not been implemented yet. They will be implemented in the next PR.
While the final pageable API will look similar to this, this is more of a test to see how the PL works.

File: sdk/digitaltwins/azure-digitaltwins-core/src/main/java/com/azure/digitaltwins/core/DigitalTwinsAsyncClient.java [hunk 2]
```diff
@@ -125,22 +126,18 @@ public HttpPipeline getHttpPipeline() {
      */
     @ServiceMethod(returns = ReturnType.SINGLE)
     public Mono<Response<String>> createRelationshipWithResponse(String digitalTwinId, String relationshipId, String relationship) {
```
Comment (azabbasi) [inline sdk/digitaltwins/azure-digitaltwins-core/src/main/java/com/azure/digitaltwins/core/DigitalTwinsAsyncClient.java:128]: won't we be having the same issue with the headers if we use the default Response type?

File: sdk/digitaltwins/azure-digitaltwins-core/src/main/java/com/azure/digitaltwins/core/implementation/serializer/DigitalTwinsStringSerializer.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+import java.io.IOException;
+
+/**
+ * Configures the {@link StdSerializer<String>} with custom behavior needed to work with Digital Twins REST API.
+ */
+public class DigitalTwinsStringSerializer extends StdSerializer<String> {
+    private static final long serialVersionUID = 1L;
+    private final ObjectMapper mapper;
+
+    public DigitalTwinsStringSerializer(Class<?> t, boolean dummy, ObjectMapper mapper) {
```
Comment (azabbasi) [inline sdk/digitaltwins/azure-digitaltwins-core/src/main/java/com/azure/digitaltwins/core/implementation/serializer/DigitalTwinsStringSerializer.java:20]: I think we need to rename this variable ... fake, mock or something

File: sdk/digitaltwins/azure-digitaltwins-core/src/main/java/com/azure/digitaltwins/core/DigitalTwinsAsyncClient.java (hunk 2 above)
Comment (abhipsaMisra) [reply sdk/digitaltwins/azure-digitaltwins-core/src/main/java/com/azure/digitaltwins/core/DigitalTwinsAsyncClient.java:128]: Yup, that is correct. This snippet is meant for testing, the actual impl is in #14495 .
 
The actual API surface is:
```java
public Mono<DigitalTwinsResponse<String>> createRelationshipWithResponse(String digitalTwinId, String relationshipId, String relationship) { }
```

File: sdk/digitaltwins/azure-digitaltwins-core/API design.md [hunk 4]
```diff
@@ -299,7 +299,31 @@ These APIs are invoked via DigitalTwinsAsyncClient.
...
+     * Creates a relationship on a digital twin.
+     *
+     * @param digitalTwinId The Id of the source digital twin.
+     * @param relationshipId The Id of the relationship to be created.
+     * @param relationship The relationship to be created.
+     * @param modelClass The model class to convert the relationship to.
+     * @return The relationship created.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public <T> Mono<T> createRelationship(String digitalTwinId, String relationshipId, Object relationship, Class<T> modelClass) { }
```
Comment (azabbasi) [inline sdk/digitaltwins/azure-digitaltwins-core/API design.md:314]: can we rename model class to typeClass or something so it won't be confused with the ADT model terminology?

File: sdk/digitaltwins/azure-digitaltwins-core/API design.md [hunk 5]
```diff
@@ -319,19 +343,40 @@ These APIs are invoked via DigitalTwinsAsyncClient.
...
     /**
      * Updates the properties of a relationship on a digital twin.
      *
      * @param digitalTwinId The Id of the source digital twin.
      * @param relationshipId The Id of the relationship to be updated.
      * @param relationshipUpdateOperations The application/json-patch+json operations to be performed on the specified digital twin's relationship.
-     * @param options The optional settings for this request.
      * @return An empty response.
      */
     @ServiceMethod(returns = ReturnType.SINGLE)
```
Comment (azabbasi) [inline sdk/digitaltwins/azure-digitaltwins-core/API design.md:378]: do we need another Mono<void> updateRelationship that takes a RequestOption?

File: sdk/digitaltwins/azure-digitaltwins-core/src/main/java/com/azure/digitaltwins/core/implementation/serializer/DigitalTwinsStringSerializer.java (hunk 3 above)
Comment (abhipsaMisra) [reply sdk/digitaltwins/azure-digitaltwins-core/src/main/java/com/azure/digitaltwins/core/implementation/serializer/DigitalTwinsStringSerializer.java:20]: I'll make use of the the following base class constructor instead:
```java
protected StdSerializer(Class<T> t) {
    this._handledType = t;
}
```

Weirdly enough, this is what the `StdSerializer` defines it as, which is why this signature was autogenerated:
```java
protected StdSerializer(Class<?> t, boolean dummy) {
    this._handledType = t;
}
```

File: sdk/digitaltwins/azure-digitaltwins-core/API design.md (hunk 5 above)
Comment (abhipsaMisra) [reply sdk/digitaltwins/azure-digitaltwins-core/API design.md:378]: No, we don't.

We have 2 overloads for `updateRelationship`:
1. `updateRelationship(String digitalTwinId, String relationshipId, List<Object> relationshipUpdateOperations)`
2. `updateRelationship(String digitalTwinId, String relationshipId, List<Object> relationshipUpdateOperations, RequestOptions options)`

The max overload one, with `RequestOptions` (2), will:
- async API -> return `withResponse`.
- sync API -> accept a `Context` input, and return `withResponse`.

**Pull request merged**